staging/src/k8s.io/client-go/tools/dnscache/dialer.go

```diff
@@ -0,0 +1,283 @@
+	return d(context.Background(), network, address)
+}
+
+// DialContext connects to the address on the named network using the provided context.
+func (d DialContext) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	return d(ctx, network, address)
+}
+
+// Stats returns stats of a dialer.
+type Stats interface {
+	Stats() DialerStats
+}
+
+// Resolver describs a net resolver. This interface is an abstract of net.Resolver.
+type Resolver interface {
+	LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error)
+}
+
+// DialerConfig contains configs for creating a dialer with dns cache.
+type DialerConfig struct {
+	// Dialer is an underlying net dialer.
+	Dialer Dialer
+	// Resolver is an underlying net resolver.
+	Resolver Resolver
+	// MinCacheDuration is the minimal cache duration for a dns result.
+	MinCacheDuration time.Duration
+	// MaxCacheDuration is the maximum cache duration for a dns result.
+	// The ttl for a cached dns result is a random value between MinCacheDuration and MaxCacheDuration.
+	MaxCacheDuration time.Duration
+	// ForceRefreshTimes specifies the max times for a cached dns result to be accessed.
+	// If a cached dns result has been accessed ForceRefreshTimes times, this dns result will be
+	// forced to deprecate. In this case, MinCacheDuration and MaxCacheDuration are ignored.
+	// If this value is less than or equal to 0, it disables dns cache.
+	ForceRefreshTimes int64
+}
+
+// cacheItem describes a cached dns result.
+type cacheItem struct {
+	host           string
+	ips            []net.IPAddr
+	expirationTime time.Time
+	usageCount     int64
+	maxUsageCount  int64
+}
+
+func newCacheItem(host string, ips []net.IPAddr, ttl time.Duration, maxUsageCount int64) *cacheItem {
+	return &cacheItem{
+		host:           host,
+		ips:            ips,
+		expirationTime: time.Now().Add(ttl),
+		usageCount:     0,
+		maxUsageCount:  maxUsageCount,
+	}
+}
+
+// ip returns an ip and a bool value which indicates whether the cache is valid.
+func (i *cacheItem) ip() (net.IPAddr, bool) {
+	if len(i.ips) <= 0 {
+		return net.IPAddr{}, false
+	}
+	count := atomic.AddInt64(&i.usageCount, 1)
+	// DNS Round-Robin.
+	index := int((count - 1)) % len(i.ips)
+	return i.ips[index], i.maxUsageCount >= count && time.Now().Before(i.expirationTime)
+}
+
+// DialerStats contains stats of a dialer.
+type DialerStats struct {
+	TotalConn          int64
+	CacheMiss          int64
+	CacheHit           int64
+	DNSQuery           int64
+	SuccessfulDNSQuery int64
+}
+
+type dialer struct {
+	dialer            Dialer
+	resolver          Resolver
+	minCacheDuration  time.Duration
+	maxCacheDuration  time.Duration
+	forceRefreshTimes int64
+	rand              *rand.Rand
+
+	lock  sync.RWMutex
+	cache map[string]*cacheItem
+
+	chanLock        sync.Mutex
+	resolveChannels map[string]chan error
+
+	stats DialerStats
+}
+
+// NewDialer creates a dialer with dns cache.
+func NewDialer(config *DialerConfig) (Dialer, error) {
```

Comment:
any in-tree reference?

Reply:
It seems there is no in-tree interface for net.Dialer.